bolt-boost/src/server.rs

```diff
@@ -0,0 +1,588 @@
+    Ok(StatusCode::OK)
+}
+
+/// Get a header with proofs for a given slot and parent hash.
+/// Spec: <https://chainbound.github.io/bolt-docs/api/builder#get_header_with_proofs>
+#[tracing::instrument(skip_all, fields(slot = params.slot))]
+async fn get_header_with_proofs(
+    State(state): State<PbsState<BuilderState>>,
+    Path(params): Path<GetHeaderParams>,
+    req_headers: HeaderMap,
+) -> Result<impl IntoResponse, PbsClientError> {
+    let ms_into_slot = ms_into_slot(params.slot, state.config.chain);
+    let max_timeout_ms = state.pbs_config().timeout_get_header_ms.min(
+        state
+            .pbs_config()
+            .late_in_slot_time_ms
+            .saturating_sub(ms_into_slot),
+    );
+
+    if max_timeout_ms == 0 {
+        warn!(
+            ms_into_slot,
+            threshold = state.pbs_config().late_in_slot_time_ms,
+            "late in slot, skipping relay requests"
+        );
+
+        return Ok(StatusCode::NO_CONTENT.into_response());
+    }
+
+    let (_, slot_uuid) = state.get_slot_and_uuid();
+
+    // prepare headers, except for start time which is set in `send_one_get_header`
+    let mut send_headers = HeaderMap::new();
+    // TODO: error handling
+    send_headers.insert(
+        HEADER_SLOT_UUID_KEY,
+        HeaderValue::from_str(&slot_uuid.to_string()).unwrap(),
+    );
+    send_headers.insert(
+        USER_AGENT,
+        get_user_agent_with_version(&req_headers).unwrap(),
+    );
+
+    let relays = state.relays();
+    let mut handles = Vec::with_capacity(relays.len());
+    for relay in relays.iter() {
+        handles.push(send_timed_get_header(
+            params,
+            relay.clone(),
+            state.config.chain,
+            state.pbs_config(),
+            send_headers.clone(),
+            ms_into_slot,
+            max_timeout_ms,
+        ));
+    }
+
+    let results = join_all(handles).await;
+    let mut relay_bids = Vec::with_capacity(relays.len());
+    let mut hash_to_proofs = HashMap::new();
+
+    // Get and remove the constraints for this slot
+    let maybe_constraints = state.data.constraints.remove(params.slot);
+
+    for (i, res) in results.into_iter().enumerate() {
+        let relay_id = relays[i].id.as_ref();
+
+        match res {
+            Ok(Some(res)) => {
+                let root = res.data.header.message.header.transactions_root;
+
+                let start = Instant::now();
+
+                // If we have constraints to verify, do that here in order to validate the bid
+                if let Some(ref constraints) = maybe_constraints {
+                    // Verify the multiproofs and continue if not valid
+                    if let Err(e) = verify_multiproofs(constraints, &res.data.proofs, root) {
+                        error!(?e, relay_id, "Failed to verify multiproof, skipping bid");
+                        continue;
+                    }
+
+                    tracing::debug!("Verified multiproof in {:?}", start.elapsed());
```

Comment:
In a slightly more optimistic setting, isn't it better to verify multiproofs only on the winning bid?

Reply:
Proof verification is not necessarily a bottleneck (takes like 50 microseconds max). It would make it more complicated to go back and find the second best bid in case the best bid is invalid